internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,582 @@
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the state with the computed module source and version.
+	if currentApplied.moduleSource != nil {
+		state.ModuleSource = types.StringValue(*currentApplied.moduleSource)
+	} else {
+		state.ModuleSource = types.StringNull()
+	}
+	if currentApplied.moduleVersion != nil {
+		state.ModuleVersion = types.StringValue(*currentApplied.moduleVersion)
+	} else {
+		state.ModuleVersion = types.StringNull()
+	}
+
+	// TODO: Eventually, when the API and SDK support speculative plan runs with a module source,
+	// this should do a speculative plan run here to determine whether changes are needed.
+
+	// Set the refreshed state, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (t *applyModuleResource) Update(ctx context.Context,
+	req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	// Retrieve values from plan.
+	var plan ApplyModuleModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// TODO: Please note that when the API and SDK support speculative runs with a module source,
+	// this will need to look at the results from the Read method's speculative run to determine
+	// whether to do an update.  A way will have to be found to force Terraform to allow the update.
+
+	// Do the run.
+	var didRun doRunOutput
+	resp.Diagnostics.Append(t.doRun(ctx, &doRunInput{
+		model: &plan,
+	}, &didRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Capture the module version in case it changed.
+	plan.ModuleVersion = types.StringValue(didRun.moduleVersion)
+
+	// Set the response state to the fully-populated plan, with or without error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+func (t *applyModuleResource) Delete(ctx context.Context,
+	req resource.DeleteRequest, resp *resource.DeleteResponse) {
+
+	// Get the current state.
+	var state ApplyModuleModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var currentApplied appliedModuleInfo
+	resp.Diagnostics.Append(t.getCurrentApplied(ctx, state, &currentApplied)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If the module source or module version is available and differs, error out.
+	if currentApplied.moduleSource != nil {
+		if state.ModuleSource.ValueString() != *currentApplied.moduleSource {
+			resp.Diagnostics.AddError("Module source differs, cannot delete", "")
+			return
+		}
+	} else if !state.ModuleSource.IsNull() {
+		resp.Diagnostics.AddError("Module source is null but state is not null, cannot delete", "")
+		return
+	}
+	if currentApplied.moduleVersion != nil {
+		if state.ModuleVersion.ValueString() != *currentApplied.moduleVersion {
+			resp.Diagnostics.AddError("Module version differs, cannot delete", "")
+			return
+		}
+	} else if !state.ModuleVersion.IsNull() {
```

Comment:
The state value for module version should never be null since it's a computed field which will always be set when a module is applied.